### PR TITLE
Activity stream

### PR DESCRIPTION
**Pull request by Jmgr**
This PR introduces an “activity stream” that records and persists partition creation and stream deletion events.

## Context

The background to this work is that we want to manage a collection of consumers that process events arising across a large collection of streams. The set of streams is dynamic - that is, streams will be programmatically created and destroyed frequently - and many of them will also be sparsely used; they will undergo periods where there are many messages, and periods when they are idle. Typically, the set of streams that is active at any given time will be a small fraction of the total number of streams.

The idea of this PR is that a stream exists that carries meta-events for a Liftbridge cluster. The meta-events supported in this PR are stream creation and deletion events, but in a later PR we also propose to include pause and resume events, assuming we can support that functionality. One or more consumers can subscribe to events in the meta-event stream and can thereby manage the process of consuming and processing messages on the regular streams as they come and go.

In this PR the meta-event stream is termed the Activity Stream. It is disabled by default and can be activated using the Liftbridge configuration.

## Implementation

The implementation proposed in this PR uses a local Liftbridge connection created on the metadata leader when elected. The activity stream in also created at this point. The events we are interested in are then recorded (pushed) into the stream. External clients can then subscribe to this stream to get a list of all past events as well as a notification when new events occur.

## Limitations / questions

The present implementation has some limitations that we would like to discuss, principally concerning the possibility that publication to the activity stream can fail. Considering the pair of events - publication of the meta-event message and the occurrence of the meta-event itself - we wonder if that pair of events can or should occur atomically or not.

In the present code, publication to the activity stream just occurs after the completion of the meta event itself (ie stream creation or deletion). If that operation fails, then the triggering request returns an error response, even though the underlying operation in fact succeeded. This doesn't serve well the requirements either of the client that triggered the operation, or the consumer of the activity stream.

To resolve this we either attempt to make the publication and the underlying operation atomic, or we say that the underlying operation is the thing that matters and the effect of that operation continues to be the source of truth. With this approach we would publish an event on the activity stream before attempting the underlying operation. Any recipient of that event would then be obliged to verify that the corresponding underlying event had in fact successfully followed by checking for the existence of the stream. Blocking operations could be provided that deal with the potential race conditions that then arise: for example, receive the stream created meta-event, and then block (with a context and timeout) for the stream itself to be created and for its leader selection to be completed). Similarly, when receiving a stream deletion meta-event, a client could check for existence of the (expected-to-be-deleted) stream and, if it still exists, block (with a context and timeout) for the error signifying deletion of the stream.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -470,6 +493,41 @@ func (s *Server) getRaft() *raftNode {
...
+	listenAddress := s.config.GetListenAddress()
+	addrs := []string{listenAddress.Host + ":" + strconv.Itoa(s.port)}
+	var err error
+	s.activityStreamClient, err = lift.Connect(addrs)
+	if err != nil {
+		return errors.Wrap(err, "failed to connect the activity stream client")
+	}
+
+	options := []lift.StreamOption{
+		lift.Group(s.config.ActivityStream.Group),
```
Comment (tylertreat) [inline server/server.go]: What is the use case for `Group` here?

File: server/server.go [hunk 2]
```diff
@@ -470,6 +493,41 @@ func (s *Server) getRaft() *raftNode {
...
+		lift.ReplicationFactor(s.config.ActivityStream.ReplicationFactor),
+		lift.Partitions(s.config.ActivityStream.Partitions),
```
Comment (tylertreat) [inline server/server.go]: Should these be configurable? (We can always expose config for them at a future time)

File: server/server.go [hunk 3]
```diff
@@ -487,7 +545,7 @@ func (s *Server) leadershipAcquired() error {
 	s.leaderSub = sub
 
 	atomic.StoreInt64(&(s.getRaft().leader), 1)
-	return nil
```
Comment (tylertreat) [inline server/server.go]: Why remove this?

File: server/server.go [hunk 4]
```diff
@@ -470,6 +493,41 @@ func (s *Server) getRaft() *raftNode {
...
+	if !s.config.ActivityStream.Enabled {
+		return nil
+	}
+
+	// Connect a local client that will be used to publish on the activity
+	// stream
+	listenAddress := s.config.GetListenAddress()
+	addrs := []string{listenAddress.Host + ":" + strconv.Itoa(s.port)}
+	var err error
+	s.activityStreamClient, err = lift.Connect(addrs)
```
Comment (tylertreat) [inline server/server.go]: Is `createActivityStream` going to get called multiple times during the life of a server? If so, we probably only want to create this client once right? Alternatively, could we just ensure `createActivityStream` gets called only up to one time per server?

Comment (tylertreat) [reply server/server.go]: Ah, later down I see we are closing and niling out `activityStreamClient` when leadership is lost. That is probably sufficient.

File: server/server.go [hunk 5]
```diff
@@ -769,3 +834,41 @@ func (s *Server) startGoroutine(f func()) {
...
+		ctx,
+		activityStream,
+		data,
+		messageOption,
+	)
+	if err != nil {
+		return errors.Wrap(err, "failed to publish a stream event")
+	}
+
+	return nil
```
Comment (tylertreat) [inline server/server.go]:
```suggestion
	return errors.Wrap(err, "failed to publish a stream event")
```

File: server/server.go (hunk 1 above)
Comment (Jmgr) [reply server/server.go]: No particular use case, I only thought that it could be useful to forward all stream creation parameters.

Comment (Jmgr) [reply server/server.go]: Should be fixed in e4a135e6901aa6126086c69cb96bd31a6d73336a.

File: server/server.go (hunk 2 above)
Comment (Jmgr) [reply server/server.go]: I have removed them for now in e4a135e6901aa6126086c69cb96bd31a6d73336a.

File: server/server.go (hunk 3 above)
Comment (Jmgr) [reply server/server.go]: This looks like a merge that has somehow failed. Should be fixed in 0cdb37a.

File: server/server.go [hunk 6]
```diff
@@ -470,6 +493,41 @@ func (s *Server) getRaft() *raftNode {
 	return r.(*raftNode)
 }
 
+// createActivityStream creates the activity stream and connects a local client
+// that will be subscribed to it.
+func (s *Server) createActivityStream() error {
```
[the remark this answers is not in the document]
Comment (Jmgr) [reply server/server.go]: This looks like a merge that has somehow failed. Should be fixed in 0cdb37a.

File: server/config.go [hunk 7]
```diff
@@ -97,6 +99,17 @@ type ClusteringConfig struct {
 	MinISR                  int
 }
 
+// ActivityStreamConfig contains settings for controlling activity stream
+// behavior.
+type ActivityStreamConfig struct {
+	Enabled              bool
+	PublicationTimeout   time.Duration
+	PublicationAckPolicy string
```
[the remark this answers is not in the document]
Comment (Jmgr) [reply server/config.go]: Yes. Should be fixed in 0674f27.

File: server/server.go [hunk 8]
```diff
@@ -30,29 +31,41 @@ import (
 
 const stateFile = "liftbridge"
 
+const (
+	streamsSubject     = "streams"
```
[the remark this answers is not in the document]
Comment (Jmgr) [reply server/server.go]: Should be fixed in da548ef.

File: server/server.go [hunk 9]
```diff
@@ -30,29 +31,41 @@ import (
...
+	activitySubject    = "activity"
+	activityStream     = activitySubject + "_stream"
+	raftSubject        = "raft"
+	replicationSubject = "replication"
+	acksSubject        = "acks"
+	publishesSubject   = "publishes"
```
[the remark this answers is not in the document]
Comment (Jmgr) [reply server/server.go]: Should be fixed in da548ef

File: server/server.go [hunk 10]
```diff
@@ -30,29 +31,41 @@ import (
 
 const stateFile = "liftbridge"
 
+const (
+	streamsSubject     = "streams"
+	activitySubject    = "activity"
```
[the remark this answers is not in the document]
Comment (Jmgr) [reply server/server.go]: Yes that makes sense. Should be fixed in da548ef.

File: server/server.go [hunk 11]
```diff
@@ -30,29 +31,41 @@ import (
...
+	activityStream     = activitySubject + "_stream"
```
[the remark this answers is not in the document]
Comment (Jmgr) [reply server/server.go]: Should be fixed in da548ef.

File: server/config.go [hunk 12]
```diff
@@ -89,46 +93,53 @@ const (
 	configClusteringReplicaMaxIdleWait      = "clustering.replica.max.idle.wait"
 	configClusteringReplicaFetchTimeout     = "clustering.replica.fetch.timeout"
 	configClusteringMinInsyncReplicas       = "clustering.min.insync.replicas"
+
+	configActivityStreamEnabled              = "activity.stream.enabled"
+	configActivityStreamPublicationTimeout   = "activity.stream.publication.timeout"
+	configActivityStreamPublicationAckPolicy = "activity.stream.publication.ack.policy"
```
Comment (tylertreat) [inline server/config.go]: Bit of a nitpick, but I propose `publish` instead of `publication`.

Comment (Jmgr) [reply server/config.go]: Sure. Should be fixed in f0a1f13.

**Comment by Jmgr** [comment]
> @Jmgr is your thought to get this activity stream work in with the expectation that the delivery guarantees/semantics might change in the future (i.e. the discussion [here](https://github.com/liftbridge-io/liftbridge-api/pull/17#issuecomment-601864590))? As it stands, the activity events in this PR would be "best effort" with no guarantee of delivery.

We will need indeed require guarantee of delivery, so we could either merge this PR now and then create another PR afterwards to implement that using the method we will find after that discussion, or postpone merging until then.

**Pull request merged**